cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -766,111 +734,35 @@ public void stop(ProgramId programId, @Nullable String runId) throws Exception {
-          cancelledInfo.ifPresent(taskInfo -> {
-            cancelledProvisionRuns.add(activeRunId);
-            // This state check is to handle a race condition where we cancel the provision task, but not in time
-            // to prevent it from sending the PROVISIONED notification.
-
-            // If the notification was sent, but not yet consumed, we are *not* done stopping the run.
-            // We have to wait for the notification to be consumed, which will start the run, and place the controller
-            // in the runtimeService. The next time we loop, we can find it in the runtimeService and tell it to stop.
-            // If the notification was not sent, then we *are* done stopping the run.
-
-            // Therefore, if the state is CREATED, we don't remove it from the iterator so that the run will get
-            // checked again in the next loop, when we may get the controller from the runtimeService to stop it.
-
-            // No other task states have this race condition, as the PROVISIONED notification is only sent
-            // after the state transitions to CREATED. Therefore it is safe to remove the runId from the iterator,
-            // as we know we are done stopping it.
-            ProvisioningOp.Status taskState = taskInfo.getProvisioningOp().getStatus();
-            if (taskState != ProvisioningOp.Status.CREATED) {
-              iterator.remove();
-            }
-          });
-        }
-      }
-
-      if (!pendingStops.isEmpty()) {
-        // If not able to stop all of them, it means there were some runs that didn't have a runtime info and
-        // didn't have a provisioning task. This can happen if the run was already finished, or the run transitioned
-        // from the provisioning state to the starting state during this stop operation.
-        // We'll get the active runs again and filter it by the pending stops. Stop will be retried for those.
-        Set<String> finalPendingStops = pendingStops;
-
-        activeRunRecords = getActiveRuns(programId, runId).entrySet().stream()
-          .filter(e -> finalPendingStops.contains(e.getKey().getRun()))
-          .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
-        pendingStops = activeRunRecords.keySet().stream().map(ProgramRunId::getRun).collect(Collectors.toSet());
-
-        if (!pendingStops.isEmpty()) {
-          TimeUnit.MILLISECONDS.sleep(200);
-        }
+    for (Map.Entry<ProgramRunId, RunRecordDetail> activeRunRecord : activeRunRecords.entrySet()) {
+      ProgramRunId activeRunId = activeRunRecord.getKey();
+      RunRecordDetail runRecord = activeRunRecord.getValue();
+
+      // Check if the program is actually started from workflow and the workflow is running
+      if (runRecord != null && runRecord.getProperties().containsKey("workflowrunid")
+        && runRecord.getStatus().equals(ProgramRunStatus.RUNNING)) {
```

Comment:
If stop is issued on a program that is running then we are throwing a bad request exception? 🤔

Reply:
this is existing code, if somebody tries to stop a program that is part of a workflow run they get a 400 response. 

For example, if they try to stop the 'phase-1' Spark program that is part of the 'DataPipelineWorkflow', they get a 400 asking them to stop the Workflow instead.